src/backend/distributed/planner/recursive_planning.c

```diff
@@ -599,6 +640,337 @@ RecursivelyPlanNonColocatedSubqueriesInWhere(Query *query,
+		 *      approach (since we lose the opportunity to push-down the whole
+		 *      sub join tree into the workers) but is easier to implement.
+		 */
+
+		RecursivelyPlanDistributedJoinNode(((JoinExpr *) node)->larg,
+										   query, recursivePlanningContext);
+
+		RecursivelyPlanDistributedJoinNode(((JoinExpr *) node)->rarg,
+										   query, recursivePlanningContext);
+
+		return;
+	}
+
+	if (!IsA(node, RangeTblRef))
+	{
+		ereport(ERROR, (errmsg("unexpected join node type (%d)",
+							   nodeTag(node))));
+	}
+
+	RangeTblRef *rangeTableRef = (RangeTblRef *) node;
+	if (IsRTERefRecurring(rangeTableRef, query))
+	{
+		return;
+	}
+
+	RangeTblEntry *distributedRte = rt_fetch(rangeTableRef->rtindex,
+											 query->rtable);
+	if (distributedRte->rtekind == RTE_RELATION)
+	{
+		ereport(DEBUG1, (errmsg("recursively planning distributed relation %s "
+								"since it is part of a distributed join node "
+								"that is outer joined with a recurring rel",
+								GetRelationNameAndAliasName(distributedRte))));
+
+		PlannerRestrictionContext *restrictionContext =
+			GetPlannerRestrictionContext(recursivePlanningContext);
+		List *requiredAttributes =
+			RequiredAttrNumbersForRelation(distributedRte, restrictionContext);
+
+		ReplaceRTERelationWithRteSubquery(distributedRte, requiredAttributes,
+										  recursivePlanningContext);
+	}
+	else if (distributedRte->rtekind == RTE_SUBQUERY)
+	{
+		/*
+		 * We don't try logging the subquery here because RecursivelyPlanSubquery
+		 * will anyway do so if the query doesn't reference the outer query.
+		 */
+		ereport(DEBUG1, (errmsg("recursively planning the distributed subquery "
+								"since it is part of a distributed join node "
+								"that is outer joined with a recurring rel")));
+
+		/*
+		 * This was our only chance to plan given distributed subquery --known
+		 * to be part of the inner side of an outer join with a recurring rel--,
+		 * so we throw a hard error instead of a DeferredError if we can't
+		 * recursively plan given subquery.
+		 */
+		if (!RecursivelyPlanSubquery(distributedRte->subquery, recursivePlanningContext))
+		{
+			/*
+			 * RecursivelyPlanSubquery fails to plan a subquery only if it
+			 * contains references to the outer query. This means that, we can't
+			 * plan such outer joins (like <recurring LEFT OUTER distributed>)
+			 * if it's a LATERAL join where the distributed side is a subquery that
+			 * references the outer side, as in,
+			 *
+			 * SELECT * FROM reference
+			 * LEFT JOIN LATERAL
+			 * (SELECT * FROM distributed WHERE reference.b > distributed.b) q
+			 * USING (a);
+			 *
+			 * For this reason, we don't need to check such cases in
+			 * DeferredErrorIfUnsupportedRecurringTuplesJoin but instead
+			 * we capture such joins here and throw an error based on this
+			 * fact.
+			 */
+			Assert(ContainsReferencesToOuterQuery(distributedRte->subquery));
+
+			ereport(ERROR, (errmsg("cannot recursively plan the distributed "
```

Comment:
is it possible to add alias of the subquery to the error message? I think it might simplify understanding the error message

Reply:
Is possible as long as we're not inlining a cte. However, I didn't want do "only print alias if aliasname is not null" thing for an ereport call.

And at that point, I guess this would require a good chunk of changes in the test outputs.